### ISSUE TITLE
SEO Settings Field (sitemap) configurations on categories is not overwriting the settings from Content SEO category group

### ISSUE THREAD

**Issue by boldenamsterdam** [nystudio107/craft-seomatic#172]
Client is having a new issue with the generated sitemap from SEOMatic.

Because Craft CMS is not giving you the option to enable/disable categories on site/language basis. All categories are by default being displayed in the sitemap. 
So we configured the categories with the 'SEO Settings Field' so the client can determine what categories should be displayed in the sitemap on site/language basis.

This was working, until the update where this issue was being fixed: [#161](https://github.com/nystudio107/craft-seomatic/issues/161)

_Production env.:_
Craft Pro  | **3.0.15**
SEOmatic | **3.0.25**

_Staging env.:_
Craft Pro  | **3.0.15**
SEOmatic | **3.1.2**

**Comment by khalwat** [comment]
I'm not entirely sure I'm understanding the issue; can't they just turn off sitemaps for certain Category Groups for certain Sites in Content SEO -> Sitemaps?

**Comment by boldenamsterdam** [comment]
Hi @khalwat ,

We have a multi lang setup **Site lang. 1** and **Site lang. 2**

1. CategoryGroup
2. Category
3. SubCategory
- Seo Settings Field (Sitemap)
-- _Site Lang_ **1: Enabled**
-- _Site Lang_ **2: Disabled**

The categories have the SEO Settings Field so the client can specify the sitemap settings on category level. Because some of the categories should not be put/displayed in the sitemap 

At the moment you will see the **SubCategory** in the sitemap of Site Lang 1 **AND** Site Lang 2.

But we actually expected that the **SubCategory** would not be displayed in the sitemap of  **Site Lang 2**.

First this was working correctly but since that other fix, I mentioned above, it's not working as expected anymore.

**Comment by khalwat** [comment]
Right, sorry, I didn't understand the issue. Fixed in https://github.com/nystudio107/craft-seomatic/commit/cde5941cda9ff779aac1f137c0b58242bd07d38e

**Comment by boldenamsterdam** [comment]
Hi @khalwat ,

After installing the update SEOMatic 3.1.3 and clearing the cache and some testing I noticed the issue is still around.
Because I'm not getting the correct sitemap as I aspected.

I will try to give in more details a better overview of the settings that I'm using.

_Settings SEO Matic _Lang 1_ **AND** _Lang 2_:_
- Content SEO
  - Category Group 1
    - Sitemap
      - Sitemap enabled (**true**)
      - Include Images & Videos in Sitemap (**true**)
      - Include Indexable Files in Sitemap (**true**)
      - Sitemap Alt URLs (**true**)

_SEO Settings on category level:_
- **1** - Category Group 
  - **1.1** - Category
    - _SEO Settings Field (**For each language**)_
      - _Sitemap (**enabled**)_
    - **1.2.1--lang1** - Sub Category  (**enabled** in Craft)
      - _SEO Settings Field (**For each language**)_
        - _Sitemap (**enabled**)_
    - **1.2.1--lang2** - Sub Category  (**enabled** in Craft)
      - _SEO Settings Field (**For each language**)_
        - _Sitemap (**disabled**)_
    - **1.2.2--lang 1** - Sub Category (**enabled** in Craft)
      - _SEO Settings Field (**For each language**)_
        - _Sitemap (**enabled**)_
     - **1.2.2--lang 2** - Sub Category (**enabled** in Craft)
       - _SEO Settings Field (**For each language**)_
         - _Sitemap (**enabled**)_
  - Category 1.2
    - Sub Category 1.2.1 (**enabled** in Craft) 
      - etc.
    - Sub Category 1.2.2 (**enabled** in Craft) 
      - etc.


What I now expected in the generated sitemap, when I go to the **"categorygroup sitemap"** of each language:

**Category Group 1 LANG 1**
_domain.com/sitemaps/1/categorygroup/categorygroup1/1/sitemap.xml_

In this sitemap I expect to see subcategory 1.2.1 but **without** an Alt. lang. URL to subcategory 1.2.1--lang-2

**Category Group 1 LANG 2**
_domain.com/sitemaps/1/categorygroup/categorygroup1/2/sitemap.xml_
 
In sitemap Category Group (lang 2) I expect not to see subcategory 1.2.1


If things are not clear or if you need more information please let me know.

Production env.:
Craft Pro | **3.0.15**
SEOmatic | **3.0.25**

Staging env.:
Craft Pro | **3.0.16.1**
SEOmatic | **3.1.3**

**Comment by boldenamsterdam** [comment]
Hi @khalwat,

Have you been able to reproduce the issue?

We did some more investigation.

One of the things we noticed is when you enable/disable a category the cache is not getting flushed. So the sitemap is not getting updated, only when we manually trigger a "clear caches" the sitemap is getting updated and the category is getting removed or added to the sitemap.

Unfortunately manually triggering the "clear caches" did not solve the original issue.

Env.
Craft Pro | **3.0.16.1**
SEOmatic | **3.1.4**

**Comment by khalwat** [comment]
Sorry, I didn't see this because it isn't a new issue.

**Comment by khalwat** [comment]
So if I'm understanding you correctly, all of the sitemaps are correct for the categories, but you're seeing it include as alternate URLs to sections where you have sitemaps disabled? Is that correct?

**Comment by aproskam** [comment]
Hi @khalwat ,

I'm not sure, I think we are misunderstanding each other.

Let me try again.

Here you can see the sitemap for the "Category Group" in both languages live:

**Sitemap lang 1:** https://kvdl.com/sitemaps/1/categorygroup/practiceAreasExpertise/1/sitemap.xml
**Sitemap lang 2:** https://kvdl.com/sitemaps/1/categorygroup/practiceAreasExpertise/2/sitemap.xml

In "Sitemap lang 1" I expect that the following url will NOT be shown in the sitemap: _https://kvdl.com/en/practice-areas-expertise/insurance-liability/letselschade_
:
But as you can see in the screenshots it is showing the category in both sitemaps:

![schermafbeelding 2018-07-24 om 20 33 43](https://user-images.githubusercontent.com/3532723/43158958-578f4934-8f81-11e8-9c23-c286d5693258.png)
![schermafbeelding 2018-07-24 om 20 33 28](https://user-images.githubusercontent.com/3532723/43158960-57a8dcc8-8f81-11e8-934f-abac81604e25.png)

I expected not to see the category in **"sitemap lang 1"** because we used the "SEO settings field" to tell SEO Matic it should not show this specific category "letselschade" in the sitemap, by setting the sitemap setting to disabled.

**Comment by khalwat** [comment]
I see, apologies for not understanding initially.

Currently, the Sitemap settings in the SEO Settings field are apparently not being applied to the sitemap at all.

I'm on the fence about whether they should be or not, or whether they just shouldn't be available in the SEO Settings field.

I take it you're using them to selectively have entries _not_ appear in the sitemap?

**Comment by aproskam** [comment]
Hi @khalwat ,

No worries. In the end we will get there.

Yeah actually we only do this because Craft by default is not allowing you to enable/disable **categories** based on language/site, on "entry" level you can enable/disable entries on language/site and the entries you disable are not showing up in the sitemap, so we thought with the SEO Settings field we had the solution to pass the Craft problem with categories.

But if I understand you, SEO Matic is not listening to the Sitemap setting that is set in the SEO Setting field. So then it makes sense why the sitemap is not generated correctly.

Do you think we can make an exception for the categories because you can not change the enable/disable status on language level for a category? And you will not be able to exclude categories from the sitemap.

**Comment by aproskam** [comment]
Besides, is it not the whole purpose of the SEO setting field that you can override the settings from Global SEO & Content SEO so you can make exceptions?

**Comment by boldenamsterdam** [comment]
Hi @khalwat ,

Did you had some time to have a thought about this sitemap SEO Setting field?

**Comment by boldenamsterdam** [comment]
Hi @khalwat ,

We would like to give our client a status update so we can manage their exceptions on how/when we can resolve this.

So we are wondering if you are still sitting on the fence about what you should do? 

Looking forward hearing from you, thanks in advance!

**Comment by khalwat** [comment]
I should be able to have a fix in for you tomorrow, assuming the testing goes well! Thanks for your patience @boldenamsterdam

**Comment by khalwat** [comment]
So just a general comment on this as well; the sitemap is just a way to tell Google what URLs you want it to spend its crawl budget on.

What you should do, if you haven't already, is ensure that for these categories, you have **Robots** set to `none`

Even if the URLs appear in the sitemap, Google is not going to index them if you have the **Robots** set to `none`

**Comment by khalwat** [comment]
I'm still testing, but this should be fixed in https://github.com/nystudio107/craft-seomatic/commit/02830ca6dd42df430f8a93740212af89ccae64e5

If you set the version in `composer.json` to `dev-develop` you can pull it down for local testing as well.

**Comment by XhmikosR** [comment]
@khalwat: the dev-develop version results in invalid XML

```
XML Parsing Error: prefix not bound to a namespace
Location: https://foo.localhost/sitemaps/1/section/pages/1/sitemap.xml
Line Number 16, Column 5:    <image:image>
----^
```

**Comment by XhmikosR** [comment]
Also, it now seems to include images where previously it didn't. Not sure if this is a regression or intended and previously it was just wrong.

**Comment by boldenamsterdam** [comment]
Hi @khalwat  

Thanks for the update, appreciated!

We did some tests with the `dev-develop` branch locally. 
Looks like the sitemaps are generated now as we would expected 🕺🏻
We did't receive any XML Parsing Errors like the comment above from XhmikosR.

Then we have two more questions.

1.
In the `<head>` we still see the following code:

![screen shot 2018-08-10 at 10 36 53](https://user-images.githubusercontent.com/7466604/43947872-7cd5ba92-9c89-11e8-98ee-1f783f7e60aa.png)

But what we want is that the alternate "en" version is not being displayed because we already told the sitemap that it should not be there. Can we not use this value to also make sure to check if there should be an alternate displayed in the `<head>`?

2.
As you explain in your general comment about configuring the robots.
Adding entries/categories to be excluded from getting indexed in the robots, this needs to be done then on entry/category level. 
Can we do this with the SEO Settings Field --> Robots ? 
If so, where will this be written that the location should be excluded? Is this added to the robots.txt or will it be specified in the `<head>` ?

Thnx!

**Comment by XhmikosR** [comment]
I guess you need to specify an image for a subentry for the invalid XML error to show up.

**Comment by khalwat** [comment]
1) I guess I could make it a signal that if the sitemaps are turned off for a particular entry, it would also not generate the `hreflang` for it 

2) What I'm saying is whether you put something in your sitemap or not, that's not the signal that Google uses to index it. You can have things not in your sitemap, and Google will eventually crawl the pages, and index them. The sitemap is just giving Google a map what you want to make sure it doesn't miss, and of your priorities

So if what you really want is to ensure that Google doesn't index certain pages, then what I would do is set **Robots** to `none` for those pages. And yes, you can do that via the SEO Settings field.

Also excluding them from the sitemap, as you're trying to do here, is okay too, but it isn't going to stop Google from potentially indexing them. **Robots** will.

**Comment by khalwat** [comment]
> Also, it now seems to include images where previously it didn't. Not sure if this is a regression or intended and previously it was just wrong.

This sounds like you have images turned OFF in the Content SEO settings, but turned ON in the Field settings.

If you intend to have images included in a section's sitemap _at all_, you'll need to enable them in Content SEO.

**Comment by khalwat** [comment]
If the SEO Settings field for an entry has **Robots** set to `none` or has sitemaps disabled, it isn't included in the `hreflang` -> https://github.com/nystudio107/craft-seomatic/commit/873a44fe901f95aadcb2867d8598946e91c988a5

**Comment by boldenamsterdam** [comment]
@khalwat not able to test at the moment because Craft 3.0.20 is required.

I will let you know if something is wrong when we have been able to test.

**Comment by khalwat** [comment]
Yep, they should be releasing 3.0.20 today, if they do things as per their normal schedule.

**Comment by boldenamsterdam** [comment]
Hi @khalwat 

We tested today the new version.
Almost everything is now being generated as we expected.

The only place where it goes wrong is in the sitemap categoryGroup from LANG 1. 
Here we still see the `hreflang alternate` and we expect it not to see because we disabled the category from the sitemap with the `SEO Settings Field --> sitemap` and we set the **`ROBOTS`** to `none`.

![screen shot 2018-08-15 at 16 17 21](https://user-images.githubusercontent.com/7466604/44153499-aebf5752-a0a8-11e8-9822-6ac174155eaa.png)

**Comment by khalwat** [comment]
So @boldenamsterdam you're saying that you're seeing these links properly _not_ be displayed for some entries/categories where you have it them disabled via the SEO Settings Field sitemap settings, but for this one in particular, it is?

Are you able to see anything unique about that category in particular?

**Comment by boldenamsterdam** [comment]
No it's not a specific case for this specific category, it's an issue for all categories.

Let me try to explain:

**Category settings**
- **Category 1**
- - Lang 1
- - - `Seo Settings Field Sitemap `: **Disabled**
- - Lang 2 
- - - `Seo Settings Field Sitemap `: **Enabled** 
---
- **Category 2**
- - Lang 1
- - - `Seo Settings Field Sitemap `: **Enabled**
- - Lang 2 
- - - `Seo Settings Field Sitemap `: **Enabled**
---
- **Sitemap Lang 1**
- - Category 1
- - - Not displayed **like expected**
- - Category 2
- - - Displayed with `hreflang alternate` **like expected**
---
- **Sitemap Lang 2**
- - Category 1
- - - Displayed with `hreflang alternate` **NOT expected** _(category 1 is disabled in lang 1 but in this sitemap it still shows an alternate link to the lang 1 version)_
- - Category 2
- - - Displayed with `hreflang alternate` **like expected**
---

**Comment by khalwat** [comment]
Thanks, that's very helpful -- just to be clear @boldenamsterdam this is _not_ fixed in the latest `dev-develop` build?

**Comment by boldenamsterdam** [comment]
Hi @khalwat ,

Just re-tested this issue with the `dev-develop` build.

It's _not_ solved, the issue is like I described in my previous comment.

**Comment by khalwat** [comment]
Phew, this is getting complicated. Alright, I see what you're expecting to happen.

Right now the logic is that if you have sitemaps disabled for a given element, then it doesn't add the `hreflang`

This part is working. But what you're asking for is if you have it enabled for a given element, but then _disabled_ for one of its translations, that it should not include it.

**Comment by khalwat** [comment]
I realize that much of this is coming from your client; something I'd again like to reiterate is just because a URL is in a sitemap, it doesn't mean Google is going to index it.

Google takes the indexing signals from the `robots` tags/headers for a given page... and it very well may find these pages by means _other_ than the sitemap.

Still, let me see if I can get this working the way you want it to.

**Comment by khalwat** [comment]
@boldenamsterdam give this a whirl: https://github.com/nystudio107/craft-seomatic/commit/ff9500f4dd32e657e2c0aa2c8e52a583e04ad117

You'll need to pull from `dev-develop`

**Comment by khalwat** [comment]
lmk if this addresses it for you @boldenamsterdam -- I'd like to roll it into the next release.

**Comment by boldenamsterdam** [comment]
@khalwat we will be back to it next week. The person who is handling this issue is currently not available.

**Comment by khalwat** [comment]
Sounds good! I'll probably roll it out anyway, and leave this issue open pending your confirmation.

It passed my internal tests, but you never know...

**Comment by khalwat** [comment]
Any luck @boldenamsterdam ?

**Comment by boldenamsterdam** [comment]
Hi @khalwat 

We did some tests this morning. Looks like everything is working!

You can close the issue, thanks for the help!

I believe the whole issue would have been much easier to fix, maybe it would not even be an issue, if the categories just had an "enabled for site" option by default from Craft CMS.
Hopefully this will be introduced in Craft CMS 3.1

**Comment by khalwat** [comment]
I agree with you re: "enabled for site" -- but we have to play the hand we're dealt! 👍
